swift/Arrow/Sources/Arrow/ArrowCExporter.swift

```diff
@@ -0,0 +1,100 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import Foundation
+import ArrowC
+
+extension String {
+    var cstring: UnsafePointer<CChar> {
+        (self as NSString).cString(using: String.Encoding.utf8.rawValue)!
+    }
+}
+
+public class ArrowCExporter {
+    private class ExportData {
+        let id: String
+        var cArray = ArrowC.ArrowArray()
+        private let arrowData: ArrowData
+        private(set) var data = [UnsafeRawPointer?]()
+        private(set) var buffers: UnsafeMutablePointer<UnsafeRawPointer?>
+        init(_ arrowData: ArrowData) {
+            id = UUID().uuidString
+            // keep a reference to the ArrowData
+            // obj so the memory doesn't get
+            // deallocated
+            self.arrowData = arrowData
+            for arrowBuffer in arrowData.buffers {
+                data.append(arrowBuffer.rawPointer)
+            }
+
+            self.buffers = UnsafeMutablePointer(mutating: data)
+            ArrowCExporter.exportedData[id.hashValue] = self
+        }
+
+        func release() {
+            // the data associated with this export data
+            // does not need to be released as they are
+            // still associated with the ArrowBuffer
+            // and it will deallocate this memory
+            ArrowCExporter.exportedData.removeValue(forKey: id.hashValue)
+        }
+    }
+
+    private static var exportedData = [Int: ExportData]()
+
+    public init() {}
+
+    public func exportType(_ cSchema: inout ArrowC.ArrowSchema, arrowType: ArrowType, name: String = "") ->
+        Result<Bool, ArrowError> {
+        do {
+            cSchema.format = try arrowType.cDataFormatId.cstring
+            cSchema.name = name.cstring
+            cSchema.release = {data in
+                data?.pointee.release = nil
+            }
+        } catch {
+            return .failure(.unknownError("\(error)"))
+        }
+        return .success(true)
+    }
+
+    public func exportField(_ schema: inout ArrowC.ArrowSchema, field: ArrowField) ->
+        Result<Bool, ArrowError> {
+            return exportType(&schema, arrowType: field.type, name: field.name)
+    }
+
+    public func exportArray(_ cArray: inout ArrowC.ArrowArray, arrowData: ArrowData) {
+        let exportData = ExportData(arrowData)
+        cArray.buffers = exportData.buffers
+        cArray.length = Int64(arrowData.length)
+        cArray.null_count = Int64(arrowData.nullCount)
+        cArray.n_buffers = Int64(arrowData.buffers.count)
+        cArray.n_children = 0
+        cArray.children = nil
+        cArray.dictionary = nil
+        cArray.private_data =
+            UnsafeMutableRawPointer(mutating: UnsafeRawPointer(bitPattern: exportData.id.hashValue))
```

Comment:
I suppose this indirect approach is because Swift does not allow pinning the `ExportData` to a fixed memory location?

Reply:
Correct, AFAIK there isn't default support for pinning the memory.